File: java-vfs/src/main/java/emg/java/vfs/filesystem/FileSystem.java
package emg.java.vfs.filesystem;

import emg.java.vfs.commands.Command;
import emg.java.vfs.files.Directory;

import java.util.Scanner;

public class FileSystem {
    public static void main(String... args) {

        final var scanner = new Scanner(System.in);
        final var root = Directory.root();
        var state = State.apply(root, root);

        while (true) {
            state.show();
            var input = scanner.nextLine();
            state = Command.from(input).apply(state);
        }
    }
}

File: kotlin-vfs/src/main/kotlin/emg/kotlin/vfs/filesystem/FileSystem.kt
package emg.kotlin.vfs.filesystem

import emg.kotlin.vfs.commands.Command
import emg.kotlin.vfs.files.Directory
import java.util.*

class FileSystem {
    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            FileSystem().main()
        }
    }

    private val scanner = Scanner(System.`in`)
    private val root = Directory.root()
    private var state = State.apply(root, root)

    fun main() {
        while (true) {
            state.show()
            val input = scanner.nextLine()
            state = Command.from(input).apply(state)
        }
    }
}